keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,444 @@
+
+## Summary
+
+In this KEP, we are proposing a way that allows stateful workloads to failover to a different node successfully after the original node is shutdown.
+
+## Motivation
+
+The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown and handle it gracefully. However, a node shutdown action may not be detected by Kubelet's Node Shutdown Mananger, either because the command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the ShutdownGracePeriod and ShutdownGracePeriodCriticalPods are not configured properly.
+
+In this KEP, we are proposing a way to handle node shutdown cases that are not detected by Kubelet's Node Shutdown Manager.
+
+### Use Cases
+
+* User wants a stateful workload to failover to another running node in case a node is shutdown without any intervention when it's safe to do so.
+
+### Goals
+
+* Increase the availability of stateful workloads when a node is shutdown.
+* Automate self-healing for stateful workloads.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager.
+
+Existing logic:
+1. When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:
+  1. [Proposed change] When a user is sure that a node is shutdown, it will apply a `quarantine` taint on the node.
+  1.  Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.
```

Comment:
"This should happen before the pods are being evicted.

this might not be true, depending when user apply the taint.

Reply:
to avoid user mistakenly put quarantine taint, I think controller could also check node status, at least the node is no longer in ready state, then controller will take actions including deleting pods etc.

Reply:
The check for node status is there.  I'll add to the KEP.

> This should happen before the pods are being evicted.
> this might not be true, depending when user apply the taint

The logic to forcefully delete the pods are added by this KEP, so we can add logic to check the taint before doing that.

Reply:
If user noticed shutdown and added taint late (more than 5 mins later), the pods are already evicted, GC controller can still force delete the pods, right? That why I think pod force deletion might happen after eviction?

Reply:
Only when node has `quarantine` taint, pod will be forcefully deleted by the Pod GC Controller, I guess. 
```
Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. 
```